src/combinedims.jl

```diff
@@ -16,23 +16,33 @@ julia> combinedims([[1, 2], [3, 4]])
  2  4
 ```
 """
-combinedims(a::AbstractArray{<:AbstractArray{<:Any, N}, M}) where {N,M} = _combinedims(a, Val(ntuple(identity, Val(N))))
+combinedims(a::AbstractArray{<:AbstractArray{<:Any, N}, M}) where {N,M} = combinedims(a, ntuple(i -> N + i, M))
+combinedims(a::AbstractArray{<:AbstractArray{<:Any, N}, 1}, outerdim::Int) where {N} = combinedims(a, (outerdim,))
+combinedims(a::AbstractArray{<:AbstractArray{<:Any, N}, M}, outerdims::NTuple{M, Int}) where {N, M} = _combinedims(a, Val(outerdims))
 
-function _combinedims(a::AbstractArray, ::Val{dims}) where {dims}
+function _combinedims(a::AbstractArray, ::Val{outer_dims}) where {outer_dims}
     outeraxes = axes(a)
     inneraxes = _inneraxes(a)
-    newaxes = (inneraxes..., outeraxes...) # TODO support permutation
-    T = inner_eltype(a)
+    ndims_total = length(outeraxes) + length(inneraxes)
+    newaxes = _combine_tuples(ndims_total, outer_dims, outeraxes, inneraxes)
 
+    T = inner_eltype(a)
     out = similar(a, T, newaxes)
-    for i in CartesianIndices(inneraxes)
-        for j in CartesianIndices(outeraxes)
-            out[i.I..., j.I...] = a[j][i]
-        end
+    for j in CartesianIndices(outeraxes)
+        I = slice_inds(j, Val(outer_dims), Val(ndims_total))
+        out[I..., :] = a[j]  # trailing ':' is required for zero-dimensional a[j]
     end
     return out
 end
 
+@inline function _combine_tuples(n, dims, t_then, t_else)
```

Comment:
Out of curiousity - does this support constant propagation? Did the old code? I wonder if there is a change with respect to a type like `StaticArrays.MArray`?

Reply:
The `combinedims` function without dimensions does infer, with dimensions - not completely... Gives a 2-union.
See https://github.com/JuliaData/SplitApplyCombine.jl/pull/49